### PR TITLE
154 remove formatters

### LINKED ISSUE #154: Remove formatters from package dependencies
Part of https://github.com/insightsengineering/coredev-tasks/issues/447

There are two functions used from formatters
- `formatters::var_labels`
- `formatters::var_relabel`

We should create a copy of this function in `teal.widgets` and export them there.

Acceptance criteria:
- remove formatters from dependencies
- update staged.dependencies.yaml files in all affected packages

### PR DESCRIPTION
**Pull request by m7pr**
Closes #154

### CODE REVIEW

**Review by chlebowa — APPROVED** [review]

File: R/formatters_var_labels.R [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+#' @seealso [formatters_var_labels()] [formatters_var_relabel()] [`formatters_var_labels<-`]
+#'
+#' @export
+#' @return \code{x} labeled by \code{label}. Note: the exact mechanism of labeling should be
+#' considered an internal implementation detail, but the label will always be retrieved via \code{obj_label}.
+#' @examples
+#' x <- formatters_with_label(c(1, 2, 3), label = "Test")
+#' obj_label(x)
+formatters_with_label <- function(x, label) {
+  obj_label(x) <- label
```
Comment (m7pr) [inline R/formatters_var_labels.R:19]: @gogonzo do you think we also need to include obj_label from formatters
https://github.com/insightsengineering/formatters/blob/ffa09bb46eb28a9e269847b918c14c719c5d28b5/R/generics.R#L261C1-L283C2 ? Looks like it's a wrapper around S3 and S4 at the same time, where for S4 it uses standardGeneric and for S3 it just calls attr(x, "label"). I think in teal packages we focus on S3, so rewriting obj_label(x) <- label to attr(x, "label") <- label should be enough

Comment (m7pr) [reply R/formatters_var_labels.R:19]: If we do not remove `formatters` from `teal.modules.clinical` we don't need to have `formatters_with_label(` at all in here

Comment (gogonzo) [reply R/formatters_var_labels.R:19]: @m7pr function not needed then since in tmc we are going to use formatters

File: NAMESPACE [hunk 2]
```diff
@@ -77,6 +78,9 @@ export(dataset_connector)
 export(dataset_connector_file)
 export(dataset_file)
 export(example_cdisc_data)
+export(formatters_var_labels)
+export(formatters_var_relabel)
+export(formatters_with_label)
```
Comment (gogonzo) [inline NAMESPACE:83]: Why to prefix functions names by `formatters_`? `teal.data::var_labels` seems to be fine. Alternatively, we can use completely new name, for example `col_labels` or `collabels`

```suggestion
export(formatters_var_labels)
export(formatters_var_relabel)
export(formatters_with_label)
```

File: R/formatters_var_labels.R (hunk 1 above)
Comment (m7pr) [reply R/formatters_var_labels.R]: removed with b2452d1c00d4c9efa86980849578876db5669466

File: NAMESPACE (hunk 2 above)
Comment (m7pr) [reply NAMESPACE]: Hey @gogonzo I wanted to keep the name of the function as it was, but figured out there might be a clash of NAMESPACEs if somewhere someone loads both formatters and teal.data and do not prefix functions. Keeping the name the same points directly it is the same function and prefixing with `formatters_` show the origin of the function.

I would stick with `formatters_var_labels` but can accept `col_labels`. Let me know what if your gut feeling in here and I'll proceed with the change here and in other packages affected by this PR.(on of course will do the same with `formatters_var_relabel`

Comment (gogonzo) [reply NAMESPACE]: What do you think about having this long name? I don't see the sense of prefixing by `formatters_`. If prefix is to inform about origin of the code - then I think it is irrelevant. In the future function in `teal.data` or `formatters` can change and the name will be missleading. I like `col_labels`

Comment (m7pr) [reply NAMESPACE]: Good, `col_labels` it is

Comment (m7pr) [reply NAMESPACE]: a44040898d1d21d58b38878faf551aafe28b6147

File: R/formatters_var_labels.R [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+#'
+#' @param x a \code{data.frame} object
+#' @param fill boolean in case the \code{label} attribute does not exist if
+#'   \code{TRUE} the variable names is returned, otherwise \code{NA}
+#'
+#' @source This function was taken 1-1 from
+#' \href{https://cran.r-project.org/web/packages/formatters/index.html}{formatters} package, to reduce the complexity of
+#' the dependency tree.
+#'
+#' @seealso [formatters_with_label()] [col_relabel()] [`col_labels<-`]
```
Comment (gogonzo) [inline R/formatters_var_labels.R:15]: 🙅 

```suggestion
#' @seealso [formatters_with_label()] [col_relabel()] [`col_labels<-`]
```

File: R/formatters_var_labels.R [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+#' This functions sets all non-missing (non-NA) variable labels in a \code{data.frame}
+#'
+#' @inheritParams col_labels
+#' @param value new variable labels, \code{NA} removes the variable label
+#'
+#' @source This function was taken 1-1 from
+#' \href{https://cran.r-project.org/web/packages/formatters/index.html}{formatters} package, to reduce the complexity of
+#' the dependency tree.
+#'
+#' @seealso [col_labels()] [col_relabel()] [formatters_with_label()]
```
Comment (gogonzo) [inline R/formatters_var_labels.R:71]:
```suggestion
#' @seealso [col_labels()] [col_relabel()]
```

File: R/formatters_var_labels.R [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+#' @seealso [col_labels()] [formatters_with_label()] [`col_labels<-`]
+#'
+#' @export
+#'
+#' @examples
+#' x <- col_relabel(iris, Sepal.Length = "Sepal Length of iris flower")
+#' col_labels(x)
+#'
+col_relabel <- function(x, ...) {
+  # todo: make this function more readable / code easier
```
Comment (gogonzo) [inline R/formatters_var_labels.R:129]: For me readibility is not an issue here
```suggestion
```

File: R/TealDataset.R [hunk 6]
```diff
@@ -184,7 +184,7 @@ TealDataset <- R6::R6Class( # nolint
     #' Derive the column labels
     #' @return `character` vector.
     get_column_labels = function() {
-      formatters::var_labels(private$.raw_data, fill = FALSE)
+      teal.data::col_labels(private$.raw_data, fill = FALSE)
```
Comment (gogonzo) [inline R/TealDataset.R:187]: Package always prioritize functions from the package
```suggestion
      col_labels(private$.raw_data, fill = FALSE)
```

File: tests/testthat/test-TealDatasetConnector.R [hunk 7]
```diff
@@ -271,7 +271,7 @@ testthat::test_that("csv_dataset_connector attritubes", {
     RACE = c("sth1|sth2", "sth", "sth"),
     stringsAsFactors = FALSE
   )
-  formatters::var_labels(ADSL_ns) <- letters[1:4] # nolint
+  teal.data::formatters_var_labels(ADSL_ns) <- letters[1:4] # nolint
```
Comment (gogonzo) [inline tests/testthat/test-TealDatasetConnector.R:274]:
```suggestion
  teal.data::var_labels(ADSL_ns) <- letters[1:4] # nolint
```

File: tests/testthat/test-TealDatasetConnector.R [hunk 8]
```diff
@@ -396,7 +396,7 @@ testthat::test_that("fun_cdisc_dataset_connector", {
     )
     x$w <- as.numeric(rnorm(40, 0, 1))
     x$ww <- as.numeric(rnorm(40, 0, 1))
-    formatters::var_labels(x) <- c("STUDYID", "USUBJID", "z", "zz", "NAs", "w", "ww")
+    teal.data::formatters_var_labels(x) <- c("STUDYID", "USUBJID", "z", "zz", "NAs", "w", "ww")
```
Comment (gogonzo) [inline tests/testthat/test-TealDatasetConnector.R:399]:
```suggestion
    var_labels(x) <- c("STUDYID", "USUBJID", "z", "zz", "NAs", "w", "ww")
```

File: tests/testthat/test-TealDatasetConnector.R [hunk 9]
```diff
@@ -413,7 +413,7 @@ testthat::test_that("fun_cdisc_dataset_connector", {
...
```
Comment (gogonzo) [inline tests/testthat/test-TealDatasetConnector.R:416]:
```suggestion
    var_labels(x) <- c("STUDYID", "USUBJID", "z", "zz", "NAs", "w", "ww")
```

File: tests/testthat/test-TealDatasetConnector.R (hunk 7 above)
Comment (m7pr) [reply tests/testthat/test-TealDatasetConnector.R]: Hey @gogonzo GitHub is having issues today as this was update in a commit, that is not visible in PRs summary of edited files lol 1d7d88ebfd4b4334e1c3ed4fdbb0d8a9f600991f - chaning to col_labels anyway

File: tests/testthat/test-TealDatasetConnector.R [hunk 10]
```diff
@@ -288,7 +288,7 @@ testthat::test_that("csv_dataset_connector attritubes", {
   testthat::expect_null(attributes(data[[1]])$label)
 
   # we should use mutate_dataset
-  data <- (x %>% mutate_dataset("formatters::var_labels(ADSL) <- letters[1:4]"))$get_raw_data()
+  data <- (x %>% mutate_dataset("teal.data::formatters_var_labels(ADSL) <- letters[1:4]"))$get_raw_data()
```
[the remark this answers is not in the document]
Comment (m7pr) [reply tests/testthat/test-TealDatasetConnector.R]: Hey @gogonzo GitHub is having issues today as this was update in a commit, that is not visible in PRs summary of edited files lol 1d7d88ebfd4b4334e1c3ed4fdbb0d8a9f600991f - chaning to col_labels anyway

File: tests/testthat/test-TealDatasetConnector.R (hunk 8 above)
Comment (m7pr) [reply tests/testthat/test-TealDatasetConnector.R]: Hey @gogonzo GitHub is having issues today as this was update in a commit, that is not visible in PRs summary of edited files lol 1d7d88ebfd4b4334e1c3ed4fdbb0d8a9f600991f - chaning to col_labels anyway

File: tests/testthat/test-TealDatasetConnector.R (hunk 9 above)
Comment (m7pr) [reply tests/testthat/test-TealDatasetConnector.R]: Hey @gogonzo GitHub is having issues today as this was update in a commit, that is not visible in PRs summary of edited files lol 1d7d88ebfd4b4334e1c3ed4fdbb0d8a9f600991f - chaning to col_labels anyway

Comment (m7pr) [reply tests/testthat/test-TealDatasetConnector.R]: 7b7ade64166f19ae0aeb670f1b40c2b6d771a11f

File: tests/testthat/test-TealDatasetConnector.R (hunk 8 above)
Comment (m7pr) [reply tests/testthat/test-TealDatasetConnector.R]: 7b7ade64166f19ae0aeb670f1b40c2b6d771a11f

File: tests/testthat/test-TealDatasetConnector.R (hunk 10 above)
[the remark this answers is not in the document]
Comment (m7pr) [reply tests/testthat/test-TealDatasetConnector.R]: 7b7ade64166f19ae0aeb670f1b40c2b6d771a11f

File: tests/testthat/test-TealDatasetConnector.R (hunk 7 above)
Comment (m7pr) [reply tests/testthat/test-TealDatasetConnector.R]: 7b7ade64166f19ae0aeb670f1b40c2b6d771a11f

**Review by gogonzo — APPROVED** [review]
👍

**Pull request merged**